libavfilter/vf_lvpdnn.c

```diff
@@ -0,0 +1,468 @@
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+/**
+ * @file 
+ * Implements livepeer frame filter using deep convolutional networks. 
+ */
+
+#include "libavformat/avio.h"
+#include "libavutil/opt.h"
+#include "libavutil/pixdesc.h"
+#include "libavutil/avassert.h"
+#include "libavutil/imgutils.h"
+#include "libavutil/hwcontext.h"
+#include "libavutil/hwcontext_internal.h"
+#include "libswscale/swscale.h"
+#include "avfilter.h"
+#include "dnn_interface.h"
+#include "formats.h"
+#include "internal.h"
+
+typedef enum {LVPDNN_CLASSIFY, LVPDNN_ODETECT} LVPDNNType;
+
+typedef struct LVPDnnContext {
+    const AVClass *class;
+
+    int     filter_type;
+    char    *model_filename;
+    DNNBackendType backend_type;
+    char    *model_inputname;
+    char    *model_outputname;
+    int     sample_rate;
+    double  valid_threshold;
+    char    *log_filename;
+
+    DNNModule   *dnn_module;
+    DNNModel    *model;
+
+    // input & output of the model at execution time
+    DNNData input;
+    DNNData output;
+    
+    struct SwsContext   *sws_rgb_scale;
+    struct SwsContext   *sws_gray8_to_grayf32;
+
+    struct AVFrame      *swscaleframe;
+    struct AVFrame      *swframeforHW;
+
+    FILE                *logfile;
+    int                 framenum;
+    
+    struct SwsContext   *sws_grayf32_to_gray8;    
+    int                 sws_uv_height;
+
+} LVPDnnContext;
+
+#define OFFSET(x) offsetof(LVPDnnContext, x)
+#define FLAGS AV_OPT_FLAG_FILTERING_PARAM | AV_OPT_FLAG_VIDEO_PARAM
+static const AVOption lvpdnn_options[] = {
+    { "filter_type", "filter type(lvpclassify/lvpodetect)",   OFFSET(filter_type),    AV_OPT_TYPE_INT,    { .i64 = 0 },    0, 1, FLAGS, "type" },
+    { "lvpclassify",    "classify filter flag",           0,                      AV_OPT_TYPE_CONST,      { .i64 = 0 },    0, 0, FLAGS, "type" },
+    { "lvpodetect",     "detect filter flag",             0,                      AV_OPT_TYPE_CONST,      { .i64 = 1 },    0, 0, FLAGS, "type" },
+    { "dnn_backend", "DNN backend",                OFFSET(backend_type),     AV_OPT_TYPE_INT,           { .i64 = 1 },    0, 1, FLAGS, "backend" },
+    { "native",      "native backend flag",        0,                        AV_OPT_TYPE_CONST,         { .i64 = 0 },    0, 0, FLAGS, "backend" },
```

Comment:
Do we need the `native` and `tensorflow` options if we already have `dnn_backend` which can be set to 0 for native and 1 for tensorflow?

Reply:
Just to clarify -  the `native` and `tensorflow` are required even if we don't use them?

Reply:
It is a description of the filter at the source level.
That is, dnn_backend can be set to two values(0,1), here 0 is for "native" and 1 for "tensorflow".